File: ruby/144_binary_tree_preorder_traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[]}

# 给定二叉树，给出先序遍历结果的数组
# 使用栈进行非递归的遍历即可

def preorder_traversal(root)
    return [] if root.nil?
    stack, res = Array.new(), Array.new()
    stack.push(root)
    while stack.length() > 0
        cur = stack.pop()
        res.push(cur.val)
        stack.push(cur.right) if cur.right
        stack.push(cur.left) if cur.left
    end

    res
end

File: cpp/144_binary_tree_preorder_traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// 二叉树的先序遍历
// 1. 使用栈
// 2. 借助`Threaded binary tree`，利用左右子树节点为空的来指向前驱和后继节点存储遍历的信息；
class Solution {
public:
	vector<int> preorderTraversal(TreeNode *root) {
		if (!root) { return {}; }
		
		stack<TreeNode *> s; s.push(root);
		vector<int> res;

		while(!s.empty()) {
			auto node = s.top(); s.pop();
			res.push_back(node->val);
			if (node->right) { s.push(node->right); }
			if (node->left) { s.push(node->left); }
		}

		return res;
	}

    vector<int> preorderTraversalNoStack(TreeNode* root) {
		if (!root) { return {}; }
		
		TreeNode *cur = root, *pre = NULL;
		vector<int> res;
		while(cur) {
			if (!cur->left) {
				res.push_back(cur->val);
				cur = cur->right;
			} else {
				pre = cur->left;
				while (pre->right && pre->right != cur) { pre = pre->right; }
				if (!pre->right) {
					res.push_back(cur->val);
					pre->right = cur;
					cur = cur->left;
				} else {
					pre->right = NULL;
					cur = cur->right;
				}
			}
		}

		return res;
    }
};
